### DISCUSSION TITLE
How do I ensure that not only the order is respected, but the execution is performed immediately after?

### DISCUSSION THREAD

**Discussion by stefanoborini** [pytest-dev/pytest-order#79] [Q&A]
From my brief experimentation, the order(after) only ensures that a test will be performed after, not necessarily immediately after. If you have a session or module fixture whose state you are relying upon, you might end up starting the "after" test when the fixture state is no longer the one of the original test.

example:

```
def test_1(mod_fixture):
    # does something with the fixture, alters its state to StateA

def test_2(mod_fixture):
    # also does something with the fixture, alters its state to StateB

@pytest.mark.order(after="test_1")
def test_3(mod_fixture):
    # assumes it starts from StateA, but in reality fixture is in StateB
```

This results in the following execution:

```
test_x::test_1 PASSED   [ 33%]
test_x::test_2 PASSED   [ 66%]
test_x::test_3 PASSED   [100%]
```

Sure, I can rely on moving the code, but it would be better to be able to specify that I want test_3 _immediately_ after test_1.

Maybe it can already be done, but I could not find how to do so.

**Comment by mrbean-bremen** [comment]
If you want to have a clearly defined order, it is probably better to use ordering by ordinal numbers instead of before/after markers, e.g.
```py
@pytest.mark.order(0)
def test_1(mod_fixture):
    # does something with the fixture, alters its state to StateA

def test_2(mod_fixture):
    # also does something with the fixture, alters its state to StateB

@pytest.mark.order(1)
def test_3(mod_fixture):
    # assumes it starts from StateA, but in reality fixture is in StateB
```

**Reply by stefanoborini** [reply]
It does guarantee the order, but does it also guarantee consequentiality (i.e. 2 is always executed _immediately_ after 1)?

**Reply by mrbean-bremen** [reply]
Yes, it does (or should, otherwise it would be a bug). First all ordered items with a positive number are executed, then all items without an order mark (in the same order they would be ordered without the plugin), and finally the ones with a negative number.
This is also described [in the documentation](https://pytest-order.readthedocs.io/en/latest/usage.html#handling-of-unordered-tests).
